### DISCUSSION TITLE
Exporting HTML Dashboards to file and re-ingest?

### DISCUSSION THREAD

**Discussion by MrCarter2959** [EvotecIT/PSWriteHTML#317] [Q&A] [answered]
I'm using PSWriteHTML to make some reports, I have the scripts outputting the HTML files and I just recently found the Navigational Piece to make a full fledge dashboard with navigation. In the example provided in https://github.com/EvotecIT/PSWriteHTML/blob/master/Examples/Example41-NavigationTopMenu/Example41-Navigation03.ps1

In the code section 

`New-HTMLPage -Name 'Page 1' {
        New-HTMLHeader {
            New-HTMLText -Text 'In header'
        }
        New-HTMLTab -Name 'Test 1' {
            New-HTMLSection -HeaderText 'Test 0' {
                New-HTMLTable -DataTable $DataTable1 -Title 'Test2' -PagingLength 2
            }
            New-HTMLSection -HeaderText 'Test 1' {
                New-HTMLTable -DataTable $DataTable1 -Title 'Test2' -PagingLength 2
            }
        }`

Would it be possible to ingest the existing HTML file, so instead of re-creating the pages for the sub-page. 

I was thinking something like a Get-Content to fetch the HTML from the file, but I'm not excatly sure how to get the HTML into the New-HTMLPage cmdlet. I tried the Add-HTML cmdlet, but it threw a error and completed. However, the content was all jumbled and non-working

**Comment by PrzemyslawKlys** [comment, marked as answer]
I understand where you're coming from and there's a reason it was built - I guess similar to yours. While the code shown in example is mainly to show how you can build something from scratch - rebuilding existing reports all the time will be a problem and time consuming.

Basically I have lots of modules that generate reports standalone and getting them to regenerate it for large domain would take hours and be inefficient. 

So I created another module which generates me the dashboard based on existing .HTML files 

![image](https://user-images.githubusercontent.com/15063294/167670930-18fff07e-aff3-4f6b-b239-929e8d4e2e97.png)

![image](https://user-images.githubusercontent.com/15063294/167671569-9d1f45bf-6aeb-440f-b78e-dd28ac01966b.png)


And it would not use New-HTMLPage - but instead would just update links in the navigation to point to specific files. It works but once you click on the link you are pointed to a report - and you loose the navigation bar on the new page, and you would go back to the main page by just pressing back in the browser.

This works, but is fairly inefficient and not user friendly, but could fit your project.

The problem with loading existing HTML will be that the report already has all the CSS/JS in the headers/footers  so before you could push it to New-HTMLPage to reuse it - you would first need to parse it and see what's there, what's missing for adding navbar and so on. Initially i planned to use my PSParseHTML module to do that - but it's going to be problematic (although I added some code in PSWriteHTML) that would help. But then some people could want to use reports created with something else then PSWriteHTML and conflicting libraries would be even messier. 

So I had another idea that instead of reimporting HTML I would just use HTML Frames. The way it works it creates an empty New-HTMLPage which just contains Navbar and the remaining part of the page is the frame that contains report. 

```powershell
  $FullPath = [io.path]::Combine($PathToSubReports, "$PageName.html")
  New-HTMLPage -Name $MenuBuilder[$Menu][$MenuReport].Name {
         New-HTMLFrame -SourcePath $MenuBuilder[$Menu][$MenuReport].Href -Scrolling Auto -Height 1500px
   } -FilePath $FullPath
```

New-HTMLFrame can be seen in examples:

```powershell
New-HTML {
        New-HTMLSection {
            New-HTMLFrame -SourcePath "$PSSCriptRoot\GPOZaurr.html" -Scrolling Auto
        } -HeaderText 'Test'
        New-HTMLSection {
            New-HTMLFrame -SourcePath "$PSSCriptRoot\GPOZaurr.html" -Scrolling Auto -Height 1500px
        } -HeaderText 'Test'
        New-HTMLSection {
            New-HTMLFrame -SourcePath "C:\Support\GitHub\PSWriteHTML\Examples\Example-Maps\Example-Maps.html"
        } -HeaderText 'Test' -Height 100vh
    } -Online -TitleText 'Test Inline' -ShowHTML -FilePath "$PSScriptRoot\Example-InlineHTML01.html" -AddComment
```

This basically "inserts" existing HTML files as a frame, but the original needs to be there. The "Page" file is just a place holder for everything else. 

So my recommendation is - go with New-HTMLFrame - and make sure your reports from other apps/modules are "available" to the dashboard (probably in some readonly folder) to display to user when they go in thru IIS/Apache or whatever. 

I will probably release my Dashboard project that allows for just configuring stuff by telling it where the files are located - but not sure how useful it will be to others.

![image](https://user-images.githubusercontent.com/15063294/167674524-ba26f9c9-3a93-4c64-b2ce-32234340c3c4.png)

**Reply by MrCarter2959** [reply]
Thank you for the fast reply! I've been playing around with it all morning, right now I have a working copy. I'm sure it's not "best practice" though...

```powershell
New-HTML -TitleText 'DashboardName' -Author 'FirstName LastName' -AutoRefresh - {
    New-HTMLNavTop -Logo 'LOGO URL' -HomeLinkHome -LogoLinkHome {
        New-NavTopMenu -Name 'PRODUCT_NAME' -IconSolid cloud {
            New-NavLink -IconSolid Book -Name "GENERAL_REPORT_AREA" {
                New-NavLink -IconSolid Book -Name 'Page1' -InternalPageID 'Page1'
                New-NavLink -IconSolid Book -Name 'Page2' -InternalPageID 'Page2'
                New-NavLink -IconSolid Book -Name 'Page3' -InternalPageID 'Page3'
            }
        }
    } 
    New-HTMLPage -Name "Page 1" -id "1" {
        New-HTMLHeader -HTMLContent {
        Add-HTML -HTML {"<br><br><br><br>"}
        New-HTMLTab -Name "Page1_ReportName" {
            
        Add-HTMLScript -Placement Footer -FilePath "PATH_TO_EXISTING_REPORT"
            }
        }
    } -FilePath "PATH_TO_PAGE_EXPORT"
} -ShowHTML -Online -FilePath "EXPORT_OF_DASHBOARD"
```

It seems to all work, its a little slow loading but it keeps the main nav header, and lets everything work togeather. Is this a acceptable way to do it? Or should I look at a better way?

After I figure this out, I'm planning on hosting it on IIS to serve the HTML page so team members can view the reports through the our intranet

***EDIT - I see the reply isn't formatting the code correctly, so I apologize for that. I'm trying to fix it, but the insert code seems to have issues with the quotes i put into it

****EDIT 2 - I see that my inner HTML Tabs aren't working, so i guess my solution isn't working as i thought it was

**Reply by PrzemyslawKlys** [reply]
Add-HTMLScript is bad idea - use New-HTMLFrame. It will be much better.

**Reply by MrCarter2959** [reply]
Thank you! You are right, that does work ALOT better... Might be a silly question, is there a way to auto expand the HTML report inside of the New-HTMLFrame? Or would i just have to set the height to a large number?

**Reply by PrzemyslawKlys** [reply]
I believe I failed to deliver auto-height as it doesn't seem to be that simple. It's still on my to-do. What you can try doing is using 80vh, 90vh, or 95vh which are values for View-port, but in the end - just like in my examples I used 1500px or something to match my requirements.
